components/service/glean/src/main/java/mozilla/components/service/glean/storages/CustomDistributionsStorageEngine.kt

```diff
@@ -0,0 +1,384 @@
+                rangeMax = range.getLong(1),
+                histogramType = histogramType,
+                values = values,
+                sum = sum
+            )
+        }
+    }
+
+    // This is a calculated read-only property that returns the total count of accumulated values
+    val count: Long
+        get() = values.map { it.value }.sum()
+
+    // This is a helper property to build the correct identifier for the metric and allow for
+    // blank categories
+    internal val identifier: String = if (category.isEmpty()) { name } else { "$category.$name" }
+
+    // This is a list of limits for the buckets.  Instantiated lazily to ensure that the range and
+    // bucket counts are set first.
+    internal val buckets: List<Long> by lazy { getBuckets() }
+
+    /**
+     * Finds the correct bucket, using a binary search to locate the index of the
+     * bucket where the sample is bigger than or equal to the bucket limit.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun findBucket(sample: Long): Long {
+        var under = 0
+        var over = bucketCount
+        var mid: Int
+
+        do {
+            mid = under + (over - under) / 2
+            if (mid == under) {
+                break
+            }
+
+            if (buckets[mid] <= sample) {
+                under = mid
+            } else {
+                over = mid
+            }
+        } while (true)
+
+        return buckets[mid]
+    }
+
+    /**
+     * Accumulates a sample to the correct bucket.
+     * If a value doesn't exist for this bucket yet, one is created.
+     *
+     * @param sample Long value representing the sample that is being accumulated
+     */
+    internal fun accumulate(sample: Long) {
+        val limit = findBucket(sample)
+        values[limit] = (values[limit] ?: 0) + 1
+        sum += sample
+    }
+
+    /**
+     * Helper function to build the [CustomDistributionData] into a JSONObject for serialization
+     * purposes.
+     */
+    internal fun toJsonObject(): JSONObject {
+        return JSONObject(mapOf(
+            "category" to category,
+            "name" to name,
+            "bucket_count" to bucketCount,
+            "range" to JSONArray(arrayOf(rangeMin, rangeMax)),
+            "histogram_type" to histogramType.toString().toLowerCase(),
+            "values" to values.mapKeys { "${it.key}" },
+            "sum" to sum
+        ))
+    }
+
+    /**
+     * Helper function to generate the list of linear bucket min values used when accumulating
+     * to the correct buckets.
+     *
+     * @return List containing the bucket limits
+     */
+    @Suppress("MagicNumber")
+    private fun getBucketsLinear(): List<Long> {
+        // Written to match the bucket generation on legacy desktop telemetry:
+        //   https://searchfox.org/mozilla-central/source/toolkit/components/telemetry/build_scripts/mozparsers/parse_histograms.py#65
```

Comment:
nit: let's use a permalink in this comment, i.e. https://searchfox.org/mozilla-central/rev/e0b0c38ee83f99d3cf868bad525ace4a395039f1/toolkit/components/telemetry/build_scripts/mozparsers/parse_histograms.py#65